geode-core/src/main/java/org/apache/geode/cache/client/internal/PoolImpl.java

```diff
@@ -1439,12 +1447,19 @@ private RuntimeException generatePoolOrCacheCancelledException(Throwable e) {
     if (re != null) {
       return re;
     }
-    InternalCache cache = GemFireCacheImpl.getInstance();
     if (cache == null) {
       if (cacheCriterion != null) {
         return cacheCriterion.generateCancelledException(e);
       }
     } else {
+      if (cache == null && dsys != null) {
+        cache = dsys.getCache();
+        if (cache == null) {
+          throw new IllegalStateException(
+              LocalizedStrings.PoolImpl_CACHE_MUST_BE_CREATED_BEFORE_CREATING_POOL
+                  .toLocalizedString());
+        }
+      }
```

Comment:
This same block of code seems to be showing up a lot, might be worth refactoring into a helper function.